ddl/partition.go

```diff
@@ -481,6 +488,535 @@ func buildTablePartitionInfo(ctx sessionctx.Context, s *ast.PartitionOptions, tb
 	return nil
 }
 
+func isPartExprUnsigned(ctx sessionctx.Context, tbInfo *model.TableInfo) bool {
+	expr, err := expression.ParseSimpleExprWithTableInfo(ctx, tbInfo.Partition.Expr, tbInfo)
+	if err != nil {
+		return false
+	}
+	pCols := expression.ExtractColumns(expr)
+	for _, col := range pCols {
+		if mysql.HasUnsignedFlag(col.GetType().GetFlag()) {
```

Comment:
Just confirm, if there is one unsigned, then all are unsigned?

Reply:
Yes, that was my idea, i.e. if one column in the partitioning expression is unsigned, then the result of the expression should be considered to be unsigned as well.

I'm not sure either we or MySQL has defined the behaviour of this.
Some information here:
https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html
which does not specify it.
I will do some tests and report back the behaviour of MySQL and see what would be a good fit for us.

Reply:
I will move this to https://github.com/pingcap/tidb/pull/36830, since it really is a separate bug: https://github.com/pingcap/tidb/issues/35827.